cmd/singularity/instance_test.go

```diff
@@ -6,67 +6,420 @@
+	}
+	if !bytes.Equal(fileContents, output) {
+		t.Fatalf("File contents were %s, but expected %s", output, fileContents)
+	}
+	// Verify that the hostname has been set correctly.
+	output, err = execInstance(instanceName, "hostname")
+	if err != nil {
+		t.Fatalf("Error executing command on instance %s: %v", instanceName, err)
+	}
+	if !bytes.Equal([]byte(testHostname+"\n"), output) {
+		t.Fatalf("Hostname is %s, but expected %s", output, testHostname)
+	}
+	// Test that the environment has been cleared.
+	_, err = execInstance(instanceName, "cat", "/home/temp/bhushan")
+	if err == nil {
+		t.Fatal("Environment file exists, but it shouldn't.")
+	}
+	// Stop the container.
+	_, err = stopInstance(stopOpts{instance: instanceName})
+	if err != nil {
+		t.Fatalf("Failed to stop instance %s: %v", instanceName, err)
+	}
+}
+
+// Test that contain works.
+func testContain(t *testing.T) {
+	const instanceName = "testcontain"
+	const fileName = "thegreattestfile"
+	// Create a temporary directory to serve as a contain directory.
+	dir, err := ioutil.TempDir("", "TestInstance")
+	if err != nil {
+		t.Fatalf("Failed to create temporary directory: %v", err)
+	}
+	defer os.RemoveAll(dir)
+	instanceOpts := startOpts{
+		contain: true,
+		workdir: dir,
+	}
+	// Start an instance with the temporary directory as the home directory.
+	_, err = startInstance(instanceImagePath, instanceName, instanceOpts)
+	if err != nil {
+		t.Fatalf("Failed to start instance %s: %v", instanceName, err)
+	}
+	// Touch a file within /tmp.
+	_, err = execInstance(instanceName, "touch", "/tmp/"+fileName)
+	if err != nil {
+		t.Fatalf("Failed to touch a file: %v", err)
+	}
+	// Stop the container.
+	_, err = stopInstance(stopOpts{instance: instanceName})
+	if err != nil {
+		t.Fatalf("Failed to stop instance %s: %v", instanceName, err)
+	}
+	// Verify that the touched file exists outside the container.
+	if _, err = os.Stat(filepath.Join(dir, "tmp", fileName)); os.IsNotExist(err) {
+		t.Fatal("The temp file doesn't exist.")
+	}
+}
+
+// Bootstrap to run all instance tests.
+func TestInstance(t *testing.T) {
+	// Build a basic Singularity image to test instances.
+	if b, err := imageBuild(buildOpts{force: true, sandbox: false}, instanceImagePath, instanceDefinition); err != nil {
+		t.Log(string(b))
+		t.Fatalf("unexpected failure: %v", err)
+	}
+	imageVerify(t, instanceImagePath, true)
+	defer os.RemoveAll(instanceImagePath)
+	// Define and loop through tests.
+	tests := []struct {
+		name       string
+		function   func(*testing.T)
+		privileged bool
+	}{
+		{"InitialNoInstances", testNoInstances, false},
+		{"BasicEchoServer", testBasicEchoServer, false},
+		{"BasicOptions", testBasicOptions, false},
+		{"Contain", testContain, false},
+		{"CreateManyInstances", testCreateManyInstances, false},
+		{"StopAll", testStopAll, false},
+		{"FinalNoInstances", testNoInstances, false},
+	}
+	for _, currentTest := range tests {
```

Comment:
on actions test we are using `tt` instead of `currentTest` , should be changed for consistency

Reply:
Changed!